quarkus-text-reviewer (#139)
lunatech-labs/lunatech-blog: posts/2024-01-10-quarkus-openai-text-review.adoc

[source,bash]
----
Act like you are a {{$from}} and writing a {{$messageType}} to a {{$to}}.

Your goal is to proofread text, make it sound smooth and polite, and correct grammatical and punctuation errors.

Additional requirements for the text:
+++++
Be brief, use short phrases, conciseness is very important: {{$isConcise}}
Be creative, use more beautiful and elegant, upper-level English literary words, but keep the meaning the same: {{$isCreative}}
Be official, the next text will be stored as an official document or part of the documentation: {{$isOfficial}}
+++++

{{$input}}
----

In the same directory, we also have to create a `config.json` file with https://learn.microsoft.com/en-us/semantic-kernel/prompts/configure-prompts?#execution-settings-used-by-ai-models[execution setting]:
[source,json]
----
{
  "schema": 1,
  "description": "Corrects the message according to the specified parameters",
  "type": "completion",
  "completion": {
    "max_tokens": 3000,
    "temperature": 0.0,
    "top_p": 0.0,
    "presence_penalty": 0.0,
    "frequency_penalty": 0.0
  }
}

----

> The most important part is to create a Semantic function for text completion using our OpenAI client and specific model id. Also, specify the skill and function name which will refer to the name of corresponding directories where we put `config.json` and `skprompt.txt` files.

Reviewer: ```suggestion
The most important part is to create a Semantic function for text completion using our OpenAI client and specific model id. Also, specify the skill and function name by referring the name of the corresponding directories where we put `config.json` and `skprompt.txt` files.
```

import com.microsoft.semantickernel.exceptions.ConfigurationException;
import com.microsoft.semantickernel.textcompletion.CompletionSKFunction;
import jakarta.enterprise.inject.Produces;

public class CompletionFunctionFactory {

    private static final String GPT_3_5_TURBO = "gpt-3.5-turbo";
    private static final String FUNCTION_NAME = "TextReview";
    private static final String SKILL_NAME = "TextReviewSkill";

    private final OpenAIClientFactory openAIClientFactory;

    public CompletionFunctionFactory(OpenAIClientFactory openAIClientFactory) {
        this.openAIClientFactory = openAIClientFactory;
    }

    @Produces
    public CompletionSKFunction getFunction() throws ConfigurationException {
        var textCompletion = SKBuilders.chatCompletion()
                .withOpenAIClient(openAIClientFactory.getClient())
                .withModelId(GPT_3_5_TURBO)
                .build();

        var kernel = SKBuilders.kernel()
                .withDefaultAIService(textCompletion)
                .build();

        return kernel.importSkillFromResources("", SKILL_NAME, FUNCTION_NAME)
                .getFunction(FUNCTION_NAME, CompletionSKFunction.class);
    }
}

----

> TextCompletionService uses a Semantic Kernel context to populate the request prompt by data provided and get the answer synchronously by calling the block() method.

Reviewer: ```suggestion
The TextCompletionService uses a Semantic Kernel context to populate the request prompt using the data provided and fetches the answer synchronously by calling the block() method.
```

        Tittle(String tittleName) {
            this.tittleName = tittleName;
        }

        public String getTittleName() {
            return tittleName;
        }
    }

    public enum MessageType {
        EMAIL("email"),
        DM("direct message");

        private final String messageTypeName;

        MessageType(String messageTypeName) {
            this.messageTypeName = messageTypeName;
        }

        public String getMessageTypeName() {
            return messageTypeName;
        }
    }

    public record OptionalFeatures(boolean isConcise, boolean isCreative, boolean isOfficial) {
    }
}
----

To run our application navigate to the root directory and execute:
[source,bash]
----
quarkus dev
----

== Frontend
Prior to developing the new frontend application, it is necessary to delete the entire directory `src/main/webui` to eliminate the default implementation. Afterward, navigate to `file-sharing-bot/src/main` and execute:

[source,bash]
----
npx create-react-app webui --template typescript
----

Additionally, include `bootstrap`. While in `/frontend/src/main/webui`, execute:
[source,bash]
----
npm install bootstrap
----

> All needful changes will be in `src/main/webui/src/App.tsx`. We need to design an interface that includes a single input field, two radio buttons for selecting "from" and "to" options, three checkboxes for additional features, an output field, and a submit button. Upon pressing the submit button, the filled information will be sent to our backend. After receiving the response, the result will be displayed in the output field.

Reviewer: ```suggestion
All the needful changes will be done in `src/main/webui/src/App.tsx`. We need to design an interface that includes a single input field, two radio buttons for selecting "from" and "to" options, three checkboxes for additional features, an output field, and a submit button. Upon pressing the submit button, the filled information will be sent to our backend. After receiving the response, the result will be displayed in the output field.
```

            </div>
          </label>
        </div>
        <div className="mb-3">
            Text:
            <input
                type="text"
                className="form-control"
                value={formData.input}
                onChange={handleTextFieldChange}
            />
        </div>
        <div className="mb-3">
          <button type="button" className="btn btn-primary" onClick={handleSubmit} disabled={loading}>
            {loading && (
              <span className="spinner-border spinner-border-sm me-2" role="status" aria-hidden="true"></span>
            )}
            Review
          </button>

        </div>
        <div className="mb-3">
            Response Text:
                <textarea
                    value={formData.responseText}
                    rows={8}
                    readOnly
                    style={{ width: '100%', resize: 'both' }}
                    className="form-control"
                />
        </div>
      </form>
    </div>
  );
};

export default App;
----

Let's restart our application
[source,bash]
----
quarkus dev
----

Now we can open http://localhost:8080 and check what we have in the result.

> == User case

Reviewer: ```suggestion
== Use case
```
